engine/execution/checker/core.go

```diff
@@ -0,0 +1,176 @@
+package checker
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/rs/zerolog"
+
+	"github.com/onflow/flow-go/engine/execution/state"
+	"github.com/onflow/flow-go/model/flow"
+	"github.com/onflow/flow-go/state/protocol"
+	"github.com/onflow/flow-go/storage"
+)
+
+// trigger the check once every minute,
+const DefaultTimeInterval = time.Minute * 1
+
+// The checker engine checks if the execution result matches the sealed result.
+type Core struct {
+	log       zerolog.Logger
+	state     protocol.State
+	execState state.ExecutionState
+}
+
+func NewCore(
+	logger zerolog.Logger,
+	state protocol.State,
+	execState state.ExecutionState,
+) *Core {
+	e := &Core{
+		log:       logger.With().Str("engine", "checker").Logger(),
+		state:     state,
+		execState: execState,
+	}
+
+	return e
+}
+
+// The point of checker engine is to perform this following check
+func checkMyCommitWithSealedCommit(
+	executedBlock *flow.Header,
+	myCommit flow.StateCommitment,
+	sealedCommit flow.StateCommitment,
+) error {
+	if myCommit != sealedCommit {
+		// mismatch
+		return fmt.Errorf("execution result is different from the sealed result, height: %v, block_id: %v, sealed_commit: %v, my_commit: %v",
+			executedBlock.Height,
+			executedBlock.ID(),
+			sealedCommit,
+			myCommit,
+		)
+	}
+
+	// match
+	return nil
+}
+
+// runLoop runs the check every minute.
+// Why using a timer instead of listening to finalized and executed events?
+// because it's simpler as it doesn't need to subscribe to those events.
+// It also runs less checks, note: the checker doesn't need to find the
+// first mismatched block, as long as it can find a mismatch, it's good enough.
+// A timer could reduce the number of checks, as it only checks once every minute.
+func (c *Core) runLoop(ctx context.Context, tickInterval time.Duration) error {
+	ticker := time.NewTicker(tickInterval)
+	defer ticker.Stop() // critical for ticker to be garbage collected
+	for {
+		select {
+		case <-ticker.C:
+			err := c.runCheck()
+			if err != nil {
+				return err
+			}
+		case <-ctx.Done():
+			return nil
+		}
+	}
```

Comment:
It might be cleaner to do this part in the engine.go

Reply:
Good point